frame/support/src/inherent.rs

```diff
@@ -135,14 +168,53 @@ macro_rules! impl_outer_inherent {
 				result
 			}
 		}
+
+		impl $crate::traits::InherentPositionCheck<$block> for $runtime {
+			fn check_inherent_position(block: &$block) -> Result<(), u32> {
+				use $crate::inherent::ProvideInherent;
+				use $crate::traits::{IsSubType, ExtrinsicCall};
+				use $crate::sp_runtime::traits::Block as _;
+
+				let mut checking_for_inherents = true;
```

Comment:
If I have understood this correct, I would propose this as a replacement logic here which is easier to understand.

```diff
diff --git a/frame/support/src/inherent.rs b/frame/support/src/inherent.rs
index 18b5052a8..9134a9dbf 100644
--- a/frame/support/src/inherent.rs
+++ b/frame/support/src/inherent.rs
@@ -175,13 +175,14 @@ macro_rules! impl_outer_inherent {
                                use $crate::traits::{IsSubType, ExtrinsicCall};
                                use $crate::sp_runtime::traits::Block as _;

-                               let mut checking_for_inherents = true;
+                               let mut first_signed_observed = false;

                                for (i, xt) in block.extrinsics().iter().enumerate() {
                                        let is_signed = $crate::inherent::Extrinsic::is_signed(xt).unwrap_or(false);

                                        let is_inherent = if is_signed {
                                                // Signed extrinsics are not inherents.
+                                               first_signed_observed = true;
                                                false
                                        } else {
                                                let mut is_inherent = false;
@@ -196,12 +197,9 @@ macro_rules! impl_outer_inherent {
                                                is_inherent
                                        };

-                                       match (checking_for_inherents, is_inherent) {
-                                               (true, true) => (),
-                                               (true, false) => checking_for_inherents = false,
-                                               // Invalid inherent position found.
-                                               (false, true) => return Err(i as u32),
-                                               (false, false) => (),
+                                       // once see the first signed transaction, no more inherents are allowed.
+                                       if first_signed_observed && is_inherent {
+                                               return Err(i as u32)
                                        }
                                }

```

Reply:
yes I refactored using your idea